File: LeetCode/Medium/0039-combination-sum/0039-combination-sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> result = new ArrayList<>();
        dfs(result, candidates, target, new ArrayList<>(), 0);

        return result;
    }

    private void dfs(List<List<Integer>> result, int[] candidates, int target, List<Integer> path, int idx) {
        if (target < 0) return;
        if (target == 0) {
            result.add(new ArrayList<>(path));
            return;
        }
        
        for (int i = idx; i < candidates.length; i++) {
            path.add(candidates[i]);
            dfs(result, candidates, target - candidates[i], path, i);
            path.remove(path.size() - 1);
        }
    }
}

File: LeetCode/Medium/0039-combination-sum/0039-combination-sum.kt
class Solution {
    fun combinationSum(candidates: IntArray, target: Int): List<List<Int>> {
        val result = mutableListOf<List<Int>>()
        val path = mutableListOf<Int>()

        fun dfs(nTarget: Int, idx: Int) {
            if (nTarget < 0) return
            if (nTarget == 0) {
                result.add(path.toList())
                return
            }

            for (i in idx until candidates.size) {
                path.add(candidates[i])
                dfs(nTarget - candidates[i], i)
                path.removeAt(path.size - 1)
            }
        }

        dfs(target, 0)
        return result
    }
}